src/memory-analyzer/analyze_symbol.cpp

```diff
@@ -0,0 +1,476 @@
+// Copyrigth 2018 Author: Malte Mues <mail.mues@gmail.com>
+#ifdef __linux__
+
+#include "analyze_symbol.h"
+#include "gdb_api.h"
+
+#include <regex>
+
+#include <ansi-c/expr2c_class.h>
+#include <goto-programs/goto_model.h>
+#include <goto-programs/read_goto_binary.h>
+#include <util/arith_tools.h>
+#include <util/c_types.h>
+#include <util/c_types_util.h>
+#include <util/cout_message.h>
+#include <util/expr.h>
+#include <util/expr_initializer.h>
+#include <util/irep.h>
+#include <util/mp_arith.h>
+#include <util/std_code.h>
+#include <util/std_expr.h>
+#include <util/std_types.h>
+#include <util/string_constant.h>
+
+symbol_analyzert::symbol_analyzert(
+  const symbol_tablet &st,
+  gdb_apit &gdb,
+  message_handlert &handler)
+  : messaget(handler), gdb_api(gdb), ns(st), c_converter(ns), id_counter(0)
+{
+}
+
+void symbol_analyzert::analyse_symbol(const std::string &symbol_name)
+{
+  const symbolt symbol = ns.lookup(symbol_name);
+  const symbol_exprt symbol_expr(symbol.name, symbol.type);
+
+  if(is_c_char_pointer(symbol.type))
+  {
+    process_char_pointer(symbol_expr, symbol.location);
+  }
+  else if(is_void_pointer(symbol.type))
+  {
+    const exprt target_expr =
+      fill_void_pointer_with_values(symbol_expr, symbol.location);
+
+    add_assignment(symbol_expr, target_expr);
+  }
+  else if(is_pointer(symbol.type))
+  {
+    const std::string memory_location = gdb_api.get_memory(symbol_name);
+
+    if(memory_location != "0x0")
+    {
+      const exprt target_symbol = process_any_pointer_target(
+        symbol_expr, memory_location, symbol.location);
+      const address_of_exprt reference_to_target(target_symbol);
+      add_assignment(symbol_expr, reference_to_target);
+    }
+    else
+    {
+      const exprt null_ptr = zero_initializer(symbol.type, symbol.location, ns);
+      add_assignment(symbol_expr, null_ptr);
+    }
+  }
+  else
+  {
+    const typet target_type = ns.follow(symbol.type);
+    const exprt target_expr = fill_expr_with_values(
+      symbol_name,
+      zero_initializer(target_type, symbol.location, ns),
+      target_type,
+      symbol.location);
+
+    if(is_struct(target_type))
```

Comment:
Make this one `if(is_struct(ns.follow(symbol.type))` or `if(symbol.type == ID_struct || symbol.type == ID_struct_tag)`